enterprise/cmd/cody-gateway/internal/actor/dotcomuser/dotcomuser.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/completions/types"
+	"github.com/sourcegraph/sourcegraph/internal/hashutil"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// user tokens are always a prefix of 4 characters (sgp_)
+// followed by a 40-character hex-encoded SHA256 hash
+const tokenLength = 4 + 40
+
+var (
+	defaultUpdateInterval = 24 * time.Hour
+)
+
+type Source struct {
+	log    log.Logger
+	cache  httpcache.Cache // TODO: add something to regularly clean up the cache
+	dotcom graphql.Client
+}
+
+var _ actor.Source = &Source{}
+
+func NewSource(logger log.Logger, cache httpcache.Cache, dotComClient graphql.Client) *Source {
+	return &Source{
+		log:    logger.Scoped("dotcomuser", "dotcom user actor source"),
+		cache:  cache,
+		dotcom: dotComClient,
+	}
+}
+
+func (s *Source) Name() string { return string(codygateway.ActorSourceDotcomUser) }
+
+func (s *Source) Get(ctx context.Context, token string) (*actor.Actor, error) {
+	// "sgp_" is sourcegraph user token
+	if token == "" || !strings.HasPrefix(token, "sgp_") {
+		return nil, actor.ErrNotFromSource{}
+	}
+
+	if len(token) != tokenLength {
+		return nil, errors.New("invalid token format")
+	}
+	hash := hashutil.ToSHA256Bytes([]byte(token))
+	digest := hex.EncodeToString(hash)
+
+	data, hit := s.cache.Get(digest)
+	if !hit {
+		return s.fetchAndCache(ctx, token, digest)
+	}
+
+	var act *actor.Actor
+	if err := json.Unmarshal(data, &act); err != nil || act == nil {
+		s.log.Error("failed to unmarshal subscription", log.Error(err))
+
+		// Delete the corrupted record.
+		s.cache.Delete(token)
+
+		return s.fetchAndCache(ctx, token, digest)
+	}
+
+	if act.LastUpdated != nil && time.Since(*act.LastUpdated) > defaultUpdateInterval {
+		return s.fetchAndCache(ctx, token, digest)
+	}
+
+	act.Source = s
+	return act, nil
+}
+
+// fetchAndCache fetches the dotcom user data for the given user token and caches it using the cacheKey
+func (s *Source) fetchAndCache(ctx context.Context, token, cacheKey string) (*actor.Actor, error) {
+	var act *actor.Actor
+	resp, checkErr := dotcom.CheckDotcomUserAccessToken(ctx, s.dotcom, token)
+	if checkErr != nil {
+		// Generate a stateless actor so that we aren't constantly hitting the dotcom API
+		act = NewActor(s, cacheKey, dotcom.CheckDotcomUserAccessTokenDotcomDotcomQueryDotcomCodyGatewayUserByTokenDotcomCodyGatewayUser{})
+	} else {
+		act = NewActor(s, cacheKey,
+			*resp.Dotcom.DotcomCodyGatewayUserByToken)
+	}
+
+	if data, err := json.Marshal(act); err != nil {
+		s.log.Error("failed to marshal actor",
+			log.Error(err))
+	} else {
+		s.cache.Set(token, data)
+	}
+
+	if checkErr != nil {
+		return nil, errors.Wrap(checkErr, "failed to validate access token")
+	}
+	return act, nil
+}
+
+// NewActor creates an actor from Sourcegraph.com user.
+func NewActor(source *Source, cacheKey string, user dotcom.CheckDotcomUserAccessTokenDotcomDotcomQueryDotcomCodyGatewayUserByTokenDotcomCodyGatewayUser) *actor.Actor {
```

Comment:
You can use the embedded fragment type here:

```suggestion
func NewActor(source *Source, cacheKey string, user dotcom.DotcomUserState) *actor.Actor {
```